### PR TITLE
simplify running operator out of cluster

### PR DESCRIPTION
**Pull request by sthaha**
## Description

This patch simplifies running operator out of cluster by just running `./scripts/run-external.sh`

## Type of change

_What type of changes does your code introduce to the Prometheus operator? Put an `x` in the box that apply._

- [ ] `CHANGE` (fix or feature that would cause existing functionality to not work as expected)
- [ ] `FEATURE` (non-breaking change which adds functionality)
- [ ] `BUGFIX` (non-breaking change which fixes an issue)
- [ ] `ENHANCEMENT` (non-breaking change which improves existing functionality)
- [x] `NONE` (if none of the other choices apply. Example, tooling, build system, CI, docs, etc.)

A sample run will look like the following 

```
❯ ./scripts/run-external.sh -c

  🔆🔆🔆  Set Cluster  🔆🔆🔆 
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
 🔔 Using cluster - 'kind-po'


  🔆🔆🔆  Ensure no other prometheus-operator is running  🔆🔆🔆 
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
 ✅ No pods with matching label app.kubernetes.io/name=prometheus-operator running in cluster


  🔆🔆🔆  Building Operator  🔆🔆🔆 
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
 ❯ make operator 
GOOS=linux GOARCH=amd64 CGO_ENABLED=0 go build -ldflags="-s -X github.com/prometheus/common/version.Revision=2117ec9b0 -X github.com/prometheus/common/version.BuildUser=sthaha -X github.com/prometheus/common/version.BuildDate=20230319-13:17:49 -X github.com/prometheus/common/version.Branch=simplify-running-op -X github.com/prometheus/common/version.Version=0.63.0" -o operator cmd/operator/main.go


  🔆🔆🔆  Update Operator CRDs  🔆🔆🔆 
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
 ❯ kubectl apply --server-side --force-conflicts -f example/prometheus-operator-crd-full 
customresourcedefinition.apiextensions.k8s.io/alertmanagerconfigs.monitoring.coreos.com serverside-applied
customresourcedefinition.apiextensions.k8s.io/alertmanagers.monitoring.coreos.com serverside-applied
customresourcedefinition.apiextensions.k8s.io/podmonitors.monitoring.coreos.com serverside-applied
customresourcedefinition.apiextensions.k8s.io/probes.monitoring.coreos.com serverside-applied
customresourcedefinition.apiextensions.k8s.io/prometheuses.monitoring.coreos.com serverside-applied
customresourcedefinition.apiextensions.k8s.io/prometheusrules.monitoring.coreos.com serverside-applied
customresourcedefinition.apiextensions.k8s.io/servicemonitors.monitoring.coreos.com serverside-applied
customresourcedefinition.apiextensions.k8s.io/thanosrulers.monitoring.coreos.com serverside-applied
 🔔 Wait for CRDs to be registered
 ❯ kubectl wait --for=condition=Established crds --all --timeout=120s 
customresourcedefinition.apiextensions.k8s.io/alertmanagerconfigs.monitoring.coreos.com condition met
customresourcedefinition.apiextensions.k8s.io/alertmanagers.monitoring.coreos.com condition met
customresourcedefinition.apiextensions.k8s.io/podmonitors.monitoring.coreos.com condition met
customresourcedefinition.apiextensions.k8s.io/probes.monitoring.coreos.com condition met
customresourcedefinition.apiextensions.k8s.io/prometheuses.monitoring.coreos.com condition met
customresourcedefinition.apiextensions.k8s.io/prometheusrules.monitoring.coreos.com condition met
customresourcedefinition.apiextensions.k8s.io/servicemonitors.monitoring.coreos.com condition met
customresourcedefinition.apiextensions.k8s.io/thanosrulers.monitoring.coreos.com condition met


  🔆🔆🔆  Run Operator  🔆🔆🔆 
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
 🔔 Running operator against cluster: kind-po - https://127.0.0.1:33169
──────────────────────────────────────────────────────────────────
 ❯ ./operator --apiserver=https://127.0.0.1:33169 --ca-file=tmp/CA_FILE --cert-file=tmp/CERT_FILE --key-file=tmp/KEY_FILE 
level=info ts=2023-03-19T03:17:54.392054466Z caller=main.go:220 msg="Starting Prometheus Operator" version="(version=0.63.0, branch=simplify-running-op, revision=2117ec9b0)"
level=info ts=2023-03-19T03:17:54.392097472Z caller=main.go:221 build_context="(go=go1.19, platform=linux/amd64, user=sthaha, date=20230319-13:17:49, tags=unknown)"
```

### CODE REVIEW

File: scripts/run-external.sh [hunk 1]
```diff
@@ -6,15 +6,49 @@ set -o pipefail
...
+	cd $(git rev-parse --show-toplevel)
+
+	local cluster="${1:-}"
+	if [[ -z "$cluster" ]]; then
+		echo "ERROR: missing cluster name"
+		echo "usage: ❯ $0 <clustername>"
+		exit 1
+	fi
+
+	run make operator
```
Comment (sthaha) [inline scripts/run-external.sh:49]: I was also thinking of adding `ensure_po_non_installed` to check that PO isn't running in cluster already.

Comment (JoaoBraveCoding) [reply scripts/run-external.sh:49]: That seems like a good idea, it it's not too complicated

**Review by JoaoBraveCoding — APPROVED** [review]
Looks great! Regardless if we add `ensure_po_non_installed` or not
Thanks for the contribution 🥳

File: scripts/run-external.sh (hunk 1 above)
Comment (simonpasquier) [reply scripts/run-external.sh:49]: I'm not sure how you could check this.

File: scripts/run-external.sh [hunk 2]
```diff
@@ -6,15 +6,49 @@ set -o pipefail
...
-
-./operator --apiserver="${apiserver}" --ca-file="${cafile}" --cert-file="${certfile}" --key-file="${keyfile}"
+run() {
+	echo " ❯ $* "
+	"$@"
+}
+
+update_crds() {
+	kubectl replace -f example/prometheus-operator-crd-full ||
+		kubectl create -f example/prometheus-operator-crd-full
```
Comment (simonpasquier) [inline scripts/run-external.sh:16]: I'd usually do
````
kubectl apply --server-side --force-conflicts -f example/prometheus-operator-crd-full
````

Comment (sthaha) [reply scripts/run-external.sh:16]: Thanks, will change to that. I thought `apply` might sometime fail if the annotations it adds is huge. Didn't know about `--server-side`

File: scripts/run-external.sh (hunk 1 above)
Comment (sthaha) [reply scripts/run-external.sh:49]: Heuristics, ... is what I had in mind .. but it may be easier to `kubectl delete -f bundle.yaml` anyway.

Comment (sthaha) [reply scripts/run-external.sh:155]: I think it may be sufficient to warn user if a deployment already exists and an option to override that check.

**Review by JoaoBraveCoding — APPROVED** [review]
lgtm :shipit:

File: scripts/run-external.sh [hunk 3]
```diff
@@ -1,20 +1,195 @@
...
+			SHOW_USAGE=true
+			# exit the loop
+			break
+			;;
+		--no-operator-run-check | -f)
+			shift
+			SKIP_OPERATOR_RUN_CHECK=true
+			shift
+			;;
+		--use-default-context | -c)
```
Comment (sthaha) [inline scripts/run-external.sh:155]: @simonpasquier @JoaoBraveCoding this part is new 
./script/run-external.sh -c   # runs against the current context.

File: scripts/run-external.sh [hunk 4]
```diff
@@ -1,20 +1,198 @@
...
+	local api_server ca_file
+
+	local cluster=".clusters[?(@.name == \"$CLUSTER\")].cluster"
+	api_server=$(kubectl config view -o jsonpath="{$cluster.server}")
+	ca_file=$(kubectl config view --raw -o jsonpath="{$cluster.certificate-authority-data}")
+
+	local key_file cert_file
+	local user=".users[?(@.name == \"$CLUSTER\")].user"
+	key_file=$(kubectl config view -o jsonpath="{$user.client-key-data}")
+	cert_file=$(kubectl config view --raw -o jsonpath="{$user.client-certificate-data}")
```
Comment (sthaha) [inline scripts/run-external.sh:63]: should these be stored into a file? `tmp/..` ?

File: scripts/run-external.sh [hunk 5]
```diff
@@ -1,20 +1,160 @@
...
+
+		Options:
+		⎻⎻⎻⎻⎻⎻⎻⎻⎻⎻⎻⎻⎻⎻⎻⎻⎻⎻⎻⎻⎻⎻⎻⎻⎻⎻⎻⎻⎻⎻⎻⎻⎻⎻⎻⎻⎻⎻⎻⎻⎻⎻⎻⎻
+		  --help | -h                   show this help
+		  --no-operator-run-check | -f: skips the check that ensures no prometheus-operator
+		                                is running in the cluster.
+
+		⎻⎻⎻⎻⎻⎻⎻⎻⎻⎻⎻⎻⎻⎻⎻⎻⎻⎻⎻⎻⎻⎻⎻⎻⎻⎻⎻⎻⎻⎻⎻⎻⎻⎻⎻⎻⎻⎻⎻⎻⎻⎻⎻⎻
+		💡 To get all cluster names in current context, run:
+		   ❯ kubectl config get-context -o name
```
[the remark this answers is not in the document]
Comment (sthaha) [reply scripts/run-external.sh]: Thanks and done.! I this gave me an idea to introduce `-c | --use-default-context` :)

File: scripts/run-external.sh [hunk 6]
```diff
@@ -1,20 +1,160 @@
...
+
+ensure_operator_not_running() {
+	$SKIP_OPERATOR_RUN_CHECK && {
+		info "skipping operator run check"
+		return 0
+	}
+
+	local po_deploy='prometheus-operator'
+
+	kubectl get deployment -A | grep -q $po_deploy && {
```
[the remark this answers is not in the document]
Comment (sthaha) [reply scripts/run-external.sh]: Done :)

**Comment by slashpai** [comment]
This change is very cool :)
/lgtm

**Review by JoaoBraveCoding — APPROVED** [review]
From my side lgtm 👍 you're a script wizard @sthaha 🧙

**Review by simonpasquier — COMMENTED** [review]
I'll check locally but lgtm :)

**Comment by simonpasquier** [comment]
Hmm i can't make it work with my kind setup

```
  🔆🔆🔆  Run Operator  🔆🔆🔆 
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
 🔔 Running operator against cluster: kind-kube-prometheus - https://127.0.0.1:43033
──────────────────────────────────────────────────────────────────
 ❯ ./operator --apiserver=https://127.0.0.1:43033 --ca-file=tmp/CA_FILE --cert-file=tmp/CERT_FILE --key-file=tmp/KEY_FILE 
level=info ts=2023-03-21T10:58:00.265850176Z caller=main.go:220 msg="Starting Prometheus Operator" version="(version=0.63.0, branch=pr/5415, revision=f69648324)"
level=info ts=2023-03-21T10:58:00.265877502Z caller=main.go:221 build_context="(go=go1.20.1, platform=linux/amd64, user=simon, date=20230321-11:57:56, tags=unknown)"
instantiating prometheus controller failed: instantiating kubernetes client failed: unable to load root certificates: unable to parse bytes as PEM block
```

**Comment by ArthurSens** [comment]
> Hmm i can't make it work with my kind setup
> 
> ```
>   🔆🔆🔆  Run Operator  🔆🔆🔆 
> ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
>  🔔 Running operator against cluster: kind-kube-prometheus - https://127.0.0.1:43033
> ──────────────────────────────────────────────────────────────────
>  ❯ ./operator --apiserver=https://127.0.0.1:43033 --ca-file=tmp/CA_FILE --cert-file=tmp/CERT_FILE --key-file=tmp/KEY_FILE 
> level=info ts=2023-03-21T10:58:00.265850176Z caller=main.go:220 msg="Starting Prometheus Operator" version="(version=0.63.0, branch=pr/5415, revision=f69648324)"
> level=info ts=2023-03-21T10:58:00.265877502Z caller=main.go:221 build_context="(go=go1.20.1, platform=linux/amd64, user=simon, date=20230321-11:57:56, tags=unknown)"
> instantiating prometheus controller failed: instantiating kubernetes client failed: unable to load root certificates: unable to parse bytes as PEM block
> ```

Totally off-topic and I don't mean to diverge the discussion from this PR, but using [Codespaces](https://github.com/features/codespaces)/[Gitpod](https://www.gitpod.io/cde) would make things like this a lot easier since everyone will have the "same machine" 😅

**Comment by sthaha** [comment]
> Hmm i can't make it work with my kind setup

@simonpasquier , thank for testing this. Could you please share more details about the kind setup? Esp .. the 
* version
* config.yaml (if any)

and How you ran the script? Did you use `-c` ?

File: scripts/run-external.sh [hunk 7]
```diff
@@ -1,20 +1,273 @@
...
+}
+
+extract_certs() {
+	local cluster=".clusters[?(@.name == \"$CLUSTER\")].cluster"
+	local user=".users[?(@.name == \"$CLUSTER\")].user"
+
+	local ca key cert
+
+	API_SERVER=$(kubectl config view -o jsonpath="{$cluster.server}")
+	ca=$(kubectl config view --raw -o jsonpath="{$cluster.certificate-authority-data}")
```
Comment (simonpasquier) [inline scripts/run-external.sh:88]: I realize that the certificate data needs to be base64 decoded. After piping the output to the `base64 -d` command, the script works for me :)

```suggestion
	ca=$(kubectl config view --raw -o jsonpath="{$cluster.certificate-authority-data} | base64 -d")
```

File: scripts/run-external.sh [hunk 8]
```diff
@@ -1,20 +1,273 @@
...
+}
+
+extract_certs() {
+	local cluster=".clusters[?(@.name == \"$CLUSTER\")].cluster"
+	local user=".users[?(@.name == \"$CLUSTER\")].user"
+
+	local ca key cert
+
+	API_SERVER=$(kubectl config view -o jsonpath="{$cluster.server}")
+	ca=$(kubectl config view --raw -o jsonpath="{$cluster.certificate-authority-data} | base64 -d")
```
Comment (simonpasquier) [inline scripts/run-external.sh:88]: sorry my initial suggestion had a typo
```suggestion
	ca=$(kubectl config view --raw -o jsonpath="{$cluster.certificate-authority-data}" | base64 -d)
```

File: scripts/run-external.sh [hunk 9]
```diff
@@ -1,20 +1,273 @@
...
+	key=$(kubectl config view --raw -o jsonpath="{$user.client-key-data}")
```
Comment (simonpasquier) [inline scripts/run-external.sh:89]:
```suggestion
	key=$(kubectl config view --raw -o jsonpath="{$user.client-key-data}" | base64 -d)
```

**Pull request merged**